common/lib/xmodule/xmodule/modulestore/tests/django_utils.py

```diff
@@ -153,40 +167,97 @@ def drop_mongo_collections(mock_create):
-TEST_DATA_SPLIT_MODULESTORE = mixed_store_config(
+TEST_DATA_SPLIT_MODULESTORE = functools.partial(
+    mixed_store_config,
     mkdtemp_clean(),
     {},
     store_order=[StoreConstructors.split, StoreConstructors.draft]
 )
 
 
-def clear_all_caches():
-    """Clear all caches so that cache info doesn't leak across test cases."""
-    # This will no longer be necessary when Django adds (in Django 1.10?):
-    #     https://code.djangoproject.com/ticket/11505
-    for cache in django.core.cache.caches.all():
-        cache.clear()
+class ModuleStoreIsolationMixin(CacheIsolationMixin):
+    """
+    A mixin to be used by TestCases that want to isolate their use of the
+    Modulestore.
+
+    How to use::
+
+        class MyTestCase(ModuleStoreMixin, TestCase):
+
+            MODULESTORE = <settings for the modulestore to test>
+
+            def my_test(self):
+                self.start_modulestore_isolation()
+                self.addCleanup(self.end_modulestore_isolation)
+
+                modulestore.create_course(...)
+                ...
+
+    """
+
+    MODULESTORE = functools.partial(mixed_store_config, mkdtemp_clean(), {})
+    CONTENTSTORE = functools.partial(contentstore_config)
+    ENABLED_CACHES = ['mongo_metadata_inheritance', 'loc_cache']
+    __settings_overrides = []
+    __old_modulestores = []
+    __old_contentstores = []
+
+    @classmethod
+    def start_modulestore_isolation(cls):
+        """
+        Isolate uses of the modulestore after this call. Once
+        :py:meth:`end_modulestore_isolation` is called, this modulestore will
+        be flushed (all content will be deleted).
+        """
+        cls.start_cache_isolation()
+        override = override_settings(
+            MODULESTORE=cls.MODULESTORE(),
+            CONTENTSTORE=cls.CONTENTSTORE(),
+        )
+
+        cls.__old_modulestores.append(copy.deepcopy(settings.MODULESTORE))
+        cls.__old_contentstores.append(copy.deepcopy(settings.CONTENTSTORE))
+        override.__enter__()
+        cls.__settings_overrides.append(override)
+        XMODULE_FACTORY_LOCK.enable()
+        clear_existing_modulestores()
+        cls.store = modulestore()
+
+    @classmethod
+    def end_modulestore_isolation(cls):
+        """
+        Delete all content in the Modulestore, and reset the Modulestore
+        settings from before :py:meth:`start_modulestore_isolation` was
+        called.
+        """
+        drop_mongo_collections()  # pylint: disable=no-value-for-parameter
+        XMODULE_FACTORY_LOCK.disable()
+        cls.__settings_overrides.pop().__exit__(None, None, None)
 
-    RequestCache().clear_request_cache()
+        assert settings.MODULESTORE == cls.__old_modulestores.pop()
+        assert settings.CONTENTSTORE == cls.__old_contentstores.pop()
+        cls.end_cache_isolation()
 
 
-class SharedModuleStoreTestCase(TestCase):
+@attr('serial')
+class SharedModuleStoreTestCase(ModuleStoreIsolationMixin, CacheIsolationTestCase):
```

Comment:
So all `SharedModuleStoreTestCase`-based tests will run serially for the time being, or is this intended to be the long-term solution?